src/NHibernate/Cache/ReadWriteCache.cs

```diff
@@ -201,21 +198,20 @@ public ISoftLock Lock(CacheKey key, object version)
 					log.Debug("Invalidating: {0}", key);
 				}
 
+				var lockValue = _cacheBase.Lock(key);
 				try
 				{
-					cache.Lock(key);
```

Comment:
The `finally` is done for unlocking, but there is nothing to unlock if the locking has failed. This line should have been before the `try`, even without having the `lockValue` to handle. All similar cases have been changed likewise.